tests/ui/needless_borrow_pat.rs

```diff
@@ -0,0 +1,109 @@
+// edition:2018
+// FIXME: run-rustfix waiting on multi-span suggestions
+
+#![warn(clippy::needless_borrow)]
+#![allow(clippy::needless_borrowed_reference)]
+
+fn f1(_: &str) {}
+macro_rules! m1 {
+    ($e:expr) => {
+        f1($e);
+    };
+}
+macro_rules! m3 {
+    ($i:ident) => {
+        Some(ref $i)
+    };
+}
+
+#[allow(dead_code)]
+fn main() {
+    let x = String::new();
+
+    // Ok, reference to a String.
+    let _: &String = match Some(x.clone()) {
+        Some(ref x) => x,
+        None => return,
+    };
+
+    // Ok, reference to a &mut String
+    let _: &&mut String = match Some(&mut x.clone()) {
+        Some(ref x) => x,
+        None => return,
+    };
+
+    // Ok, the pattern is in a different context than the match arm
```

Comment:
Maybe this is a better reason to not lint here: `pat.span.ctxt() != ident.span.ctxt()`
```suggestion
    // Ok, the pattern is in a different context than the identifier
```

Reply:
The identifier being different doesn't actually matter. Something like:

```rust
macro_rules! m {
    ($x:ident) => {
        match Some(x) {
            Some(ref $x) => *$x,
            None => return,
        }
    };
}
```

would lint fine (almost fine, the suggestion is wrong). I'm struggling to come up with a useful example where it will happen.